joint_calling/upload_processor.py

```diff
@@ -0,0 +1,110 @@
+    docker_image: str = None,
+    key: str = None,
+) -> List:
+    """This script takes a Hail.Batch workflow and a list of files to
+    move between buckets. It adds 1 job per file to the workflow, and
+    runs in a docker_image. It returns the list of created jobs. Each
+    job consists of a gsutil move command for each valid file.
+    When run, these jobs will perform the batch move file operation.
+
+    Parameters
+    ==========
+    batch: hb.Batch
+        An object representing the DAG of jobs to run.
+    files: List[str]
+        A list of the file names to be moved.
+        For example ["TOB1543.g.vcf.gz","TOB2314.g.vcf.gz","TOB3423.g.vcf.gz"]
+    source_prefix: str
+        The path to the sub-directory where the files are initially located.
+        For example "cpg-tob-wgs-upload" or "cpg-tob-wgs-upload/v1"
+    destination_prefix: str
+        The path to the sub-directory where the files should be moved.
+        For example "cpg-tob-wgs-main" or "cpg-tob-wgs-upload/batch0"
+    docker_image: str, optional
+        The address and tag of a previously built docker image, within the
+        artifact registry. Each batch job will run in this image.
+        For example;
+        australia-southeast1-docker.pkg.dev/project/images/driver:version'
+    key: str, optional
+        key-file for the service account used for authentication. In the case that this
+        is not provided as an input, it is assumed that this key will exist at
+        /gsa-key/key.json. This is the case when using the hail batch service backend.
+        For example:
+        "{
+            "type": "service_account",
+            "project_id": "",
+            "private_key_id": "" ...
+        }
+        " """
+
+    jobs = []
+
+    for sample in files:
+
+        previous_location = os.path.join('gs://', source_prefix, sample)
+        new_location = os.path.join('gs://', destination_prefix, sample)
+
+        # Checks if the files exist at the source and destination
+        get_file_source = subprocess.run(
+            ['gsutil', '-q', 'stat', previous_location], check=False
+        )
+        get_file_destination = subprocess.run(
+            ['gsutil', '-q', 'stat', new_location], check=False
+        )
+
+        # In the case that the file is not found at the source
+        if get_file_source.returncode == 1:
+            # Valid - File has already been moved in a previous run.
+            # i.e. it exists at the destination and not the source.
+            if get_file_destination.returncode == 0:
+                continue
+
+            # Invalid - the file does not exist at either location
+            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), sample)
+
+        j = batch.new_job(name=f'moving_{sample}')
+
+        if docker_image is not None:
+            j.image(docker_image)
+
+        # Authenticate to service account.
+        if key is not None:
+            j.command(f"echo '{key}' > key.json")
```

Comment:
Probably better to use an absolute path for the key here, e.g. something in the `/tmp` directory.

Reply:
Have amended this :)